template/src/ontology/run.sh.jinja2

```diff
@@ -45,21 +56,30 @@ if [ -n "$ODK_BINDS" ]; then
     VOLUME_BIND="$VOLUME_BIND,$ODK_BINDS"
 fi
 
+# Gather environment variables into a single file
+cat <<EOF > run.sh.env
+ROBOT_JAVA_ARGS=$ODK_JAVA_OPTS
+JAVA_OPTS=$ODK_JAVA_OPTS
+GH_TOKEN=$GH_TOKEN
+EOF
+
 if [ -n "$USE_SINGULARITY" ]; then
     {% if project.use_external_date is sameas True %}export TODAY=$(date +%Y-%m-%d){% endif %}
     singularity exec --cleanenv $ODK_SINGULARITY_OPTIONS \
-        --env "ROBOT_JAVA_ARGS=$ODK_JAVA_OPTS,JAVA_OPTS=$ODK_JAVA_OPTS" \
+        --env-file run.sh.env \
         --bind $VOLUME_BIND \
         -W $WORK_DIR \
         docker://obolibrary/$ODK_IMAGE:$ODK_TAG $TIMECMD "$@"
 else
     BIND_OPTIONS="-v $(echo $VOLUME_BIND | sed 's/,/ -v /')"
     docker run -u $(id -u):$(id -g) $ODK_DOCKER_OPTIONS $BIND_OPTIONS -w $WORK_DIR \
-        -e ROBOT_JAVA_ARGS="$ODK_JAVA_OPTS" -e JAVA_OPTS="$ODK_JAVA_OPTS" \
+        --env-file run.sh.env \
         {% if project.use_external_date is sameas True %}-e TODAY=`date +%Y-%m-%d` {% endif %} \
         --rm -ti obolibrary/$ODK_IMAGE:$ODK_TAG $TIMECMD "$@"
 fi
 
+rm run.sh.env
```

Comment:
What happens if you have two shells open:

1. one is running a release (long running, needs the token)
2. The other, different shell, is just checking something real quick `sh run.sh robot -i...`.

Will 2 not delete the file before 1 can use it and thus make it fail?

Reply:
As long as the second process happens _after_ the first process has reached the `docker run` command, that wouldn’t be a problem. At this point the container of the first process is already running with the variable in its environment.

There _is_ a very slight possibility for a race condition if you start the second process _immediately_ after the first, and the second process runs so fast that it terminates (and thus deletes the `run.sh.env` file) before the first process even has the time to reach the `docker run` command.

I consider that highly unlikely, but if we want to avoid even that possibility, we can give the `.env` a unique, process-dependent filename (e.g. `run.sh.env.$$`, where `$$` is the PID of the current process).

Or, we forget about the `.env` file altogether and go back to passing environment variables on the command line, if we decide that preventing the GitHub token from appearing in a command line is not worth (see previous comment).

Reply:
Ah ok, no no, I forgot that docker would evaluate the `run.sh.env` when the container is started, not like "binding a file into the container". Sounds good. Maybe add `rm -f` to avoid the minor risk of a race condition?